### DISCUSSION TITLE
Relative auto-import top level

### DISCUSSION THREAD

**Discussion by adriengcql** [microsoft/pylance-release#4802] [Enhancement]
As an enthusiastic user of the auto-import completion, there is one thing I couldn't manage to deal with: when setting the import format to relative (`"python.analysis.importFormat": "relative"`). there are still cases that need local absolute imports.

For instance in a project with multiple python packages or a separate tests folder:
```
 - myproject
       - mypackage
            - __init__.py
       -  tests
            - test_foo.py
```
Importing from `mypackage` in `test_foo.py` results in a relative import `from ..mypackage import myfunc` which is not what is wanted.

In short the idea is to have a way to switch to absolute imports in certain cases, either with some manual configuration or by detecting when the import goes beyond the root of the package.

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.

**Comment by Tinche** [comment]
I've been getting this for ages both in my work and open source codebases, both using Poetry and PDM.

If a project contains multiple packages (or just a main package and another one for tests), using relative imports in Pylance simply doesn't work properly. The generated relative import is too deep and doesn't work at runtime. This makes the auto import feature with relative imports much less useful since I need to go back to the top of the file and fix them constantly.

Here's an example of me trying to write a unit test for my cattrs library:

![Screenshot 2024-11-08 at 20 50 46](https://github.com/user-attachments/assets/43d2e1c7-6fae-4dda-aef6-e0b618ca0eaf)
![Screenshot 2024-11-08 at 20 50 51](https://github.com/user-attachments/assets/fc6acec0-3236-4199-b691-67a411f974f3)

It doesn't seem to me like there's much to discuss here.

**Comment by maxeonyx** [comment]
Linking to this which is the same issue: https://github.com/microsoft/pylance-release/discussions/6467

**Reply by maxeonyx** [reply]
If one is to be closed it should be mine. This one has more upvotes.

**Comment by npalmius** [comment]
Maybe I'm missing something, but how do I up-vote this? And why is it being considered as an enhancement? Seems like a defect to me.

**Reply by rchiodo** [reply]
You can upvote the top most item in the discussion. It's what we see here:

![image](https://github.com/user-attachments/assets/493b60ba-0b42-495f-b5c2-bca10ba4642e)

**Reply by npalmius** [reply]
Ah got it, thanks @rchiodo!

**Comment by umarbutler** [comment]
+1 This is a bug, not a feature.
